File: app/src/main/java/org/stepic/droid/storage/dao/LessonDaoImpl.java
package org.stepic.droid.storage.dao;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import org.stepic.droid.model.Lesson;
import org.stepic.droid.storage.structure.DbStructureLesson;
import org.stepic.droid.util.DbParseHelper;

import javax.inject.Inject;

public class LessonDaoImpl extends DaoBase<Lesson> {

    @Inject
    public LessonDaoImpl(SQLiteDatabase openHelper) {
        super(openHelper);
    }

    @Override
    public Lesson parsePersistentObject(Cursor cursor) {
        Lesson lesson = new Lesson();
        int columnIndexLessonId = cursor.getColumnIndex(DbStructureLesson.Column.LESSON_ID);
        int columnIndexSteps = cursor.getColumnIndex(DbStructureLesson.Column.STEPS);
        int columnIndexIsFeatured = cursor.getColumnIndex(DbStructureLesson.Column.IS_FEATURED);
        int columnIndexIsPrime = cursor.getColumnIndex(DbStructureLesson.Column.IS_PRIME);
        int columnIndexProgress = cursor.getColumnIndex(DbStructureLesson.Column.PROGRESS);
        int columnIndexOwner = cursor.getColumnIndex(DbStructureLesson.Column.OWNER);
        int columnIndexSubscriptions = cursor.getColumnIndex(DbStructureLesson.Column.SUBSCRIPTIONS);
        int columnIndexViewedBy = cursor.getColumnIndex(DbStructureLesson.Column.VIEWED_BY);
        int columnIndexPassedBy = cursor.getColumnIndex(DbStructureLesson.Column.PASSED_BY);
        int columnIndexDependencies = cursor.getColumnIndex(DbStructureLesson.Column.DEPENDENCIES);
        int columnIndexIsPublic = cursor.getColumnIndex(DbStructureLesson.Column.IS_PUBLIC);
        int columnIndexTitle = cursor.getColumnIndex(DbStructureLesson.Column.TITLE);
        int columnIndexSlug = cursor.getColumnIndex(DbStructureLesson.Column.SLUG);
        int columnIndexCreateDate = cursor.getColumnIndex(DbStructureLesson.Column.CREATE_DATE);
        int columnIndexLearnersGroup = cursor.getColumnIndex(DbStructureLesson.Column.LEARNERS_GROUP);
        int columnIndexTeacherGroup = cursor.getColumnIndex(DbStructureLesson.Column.TEACHER_GROUP);
        int indexIsCached = cursor.getColumnIndex(DbStructureLesson.Column.IS_CACHED);
        int indexIsLoading = cursor.getColumnIndex(DbStructureLesson.Column.IS_LOADING);
        int indexCoverURL = cursor.getColumnIndex(DbStructureLesson.Column.COVER_URL);

        lesson.setId(cursor.getLong(columnIndexLessonId));
        lesson.setSteps(DbParseHelper.parseStringToLongArray(cursor.getString(columnIndexSteps)));
        lesson.set_featured(cursor.getInt(columnIndexIsFeatured) > 0);
        lesson.set_prime(cursor.getInt(columnIndexIsPrime) > 0);
        lesson.setProgress(cursor.getString(columnIndexProgress));
        lesson.setOwner(cursor.getInt(columnIndexOwner));
        lesson.setSubscriptions(DbParseHelper.parseStringToStringArray(cursor.getString(columnIndexSubscriptions)));
        lesson.setViewed_by(cursor.getInt(columnIndexViewedBy));
        lesson.setPassed_by(cursor.getInt(columnIndexPassedBy));
        lesson.setDependencies(DbParseHelper.parseStringToStringArray(cursor.getString(columnIndexDependencies)));
        lesson.set_public(cursor.getInt(columnIndexIsPublic) > 0);
        lesson.setTitle(cursor.getString(columnIndexTitle));
        lesson.setSlug(cursor.getString(columnIndexSlug));
        lesson.setCreate_date(cursor.getString(columnIndexCreateDate));
        lesson.setLearners_group(cursor.getString(columnIndexLearnersGroup));
        lesson.setTeacher_group(cursor.getString(columnIndexTeacherGroup));
        lesson.set_cached(cursor.getInt(indexIsCached) > 0);
        lesson.set_loading(cursor.getInt(indexIsLoading) > 0);
        lesson.setCover_url(cursor.getString(indexCoverURL));

        return lesson;
    }

    @Override
    public String getDbName() {
        return DbStructureLesson.LESSONS;
    }

    @Override
    public ContentValues getContentValues(Lesson lesson) {
        ContentValues values = new ContentValues();

        values.put(DbStructureLesson.Column.LESSON_ID, lesson.getId());
        values.put(DbStructureLesson.Column.STEPS, DbParseHelper.parseLongArrayToString(lesson.getSteps()));
        values.put(DbStructureLesson.Column.IS_FEATURED, lesson.is_featured());
        values.put(DbStructureLesson.Column.IS_PRIME, lesson.is_prime());
        values.put(DbStructureLesson.Column.PROGRESS, lesson.getProgress());
        values.put(DbStructureLesson.Column.OWNER, lesson.getOwner());
        values.put(DbStructureLesson.Column.SUBSCRIPTIONS, DbParseHelper.parseStringArrayToString(lesson.getSubscriptions()));
        values.put(DbStructureLesson.Column.VIEWED_BY, lesson.getViewed_by());
        values.put(DbStructureLesson.Column.PASSED_BY, lesson.getPassed_by());
        values.put(DbStructureLesson.Column.DEPENDENCIES, DbParseHelper.parseStringArrayToString(lesson.getDependencies()));
        values.put(DbStructureLesson.Column.IS_PUBLIC, lesson.is_public());
        values.put(DbStructureLesson.Column.TITLE, lesson.getTitle());
        values.put(DbStructureLesson.Column.SLUG, lesson.getSlug());
        values.put(DbStructureLesson.Column.CREATE_DATE, lesson.getCreate_date());
        values.put(DbStructureLesson.Column.LEARNERS_GROUP, lesson.getLearners_group());
        values.put(DbStructureLesson.Column.TEACHER_GROUP, lesson.getTeacher_group());
        values.put(DbStructureLesson.Column.COVER_URL, lesson.getCover_url());

        return values;
    }

    @Override
    public String getDefaultPrimaryColumn() {
        return DbStructureLesson.Column.LESSON_ID;
    }

    @Override
    public String getDefaultPrimaryValue(Lesson persistentObject) {
        return persistentObject.getId()+"";
    }
}

File: app/src/main/java/org/stepic/droid/storage/dao/LessonDaoImpl.kt
package org.stepic.droid.storage.dao

import android.content.ContentValues
import android.database.Cursor
import com.google.gson.Gson
import org.stepic.droid.storage.operations.DatabaseOperations
import org.stepic.droid.util.DbParseHelper
import org.stepic.droid.util.getBoolean
import org.stepic.droid.util.getDate
import org.stepic.droid.util.getLong
import org.stepic.droid.util.getString
import org.stepic.droid.util.toObject
import org.stepik.android.cache.lesson.structure.DbStructureLesson
import org.stepik.android.model.Lesson
import javax.inject.Inject

class LessonDaoImpl
@Inject
constructor(
    databaseOperations: DatabaseOperations,
    private val gson: Gson
) : DaoBase<Lesson>(databaseOperations) {
    public override fun getDbName() =
        DbStructureLesson.TABLE_NAME

    public override fun getDefaultPrimaryColumn(): String =
        DbStructureLesson.Columns.ID

    public override fun getDefaultPrimaryValue(persistentObject: Lesson): String =
        persistentObject.id.toString()

    public override fun parsePersistentObject(cursor: Cursor): Lesson =
        Lesson(
            id = cursor.getLong(DbStructureLesson.Columns.ID),
            title = cursor.getString(DbStructureLesson.Columns.TITLE),
            slug = cursor.getString(DbStructureLesson.Columns.SLUG),
            coverUrl = cursor.getString(DbStructureLesson.Columns.COVER_URL),
            courses = DbParseHelper.parseStringToLongArray(cursor.getString(DbStructureLesson.Columns.COURSES)) ?: longArrayOf(),
            steps = DbParseHelper.parseStringToLongArray(cursor.getString(DbStructureLesson.Columns.STEPS)) ?: longArrayOf(),
            actions = cursor.getString(DbStructureLesson.Columns.ACTIONS)?.toObject(gson),
            isFeatured = cursor.getBoolean(DbStructureLesson.Columns.IS_FEATURED),
            progress = cursor.getString(DbStructureLesson.Columns.PROGRESS),
            owner = cursor.getLong(DbStructureLesson.Columns.OWNER),
            subscriptions = DbParseHelper.parseStringToStringArray(cursor.getString(DbStructureLesson.Columns.SUBSCRIPTIONS)),
            viewedBy = cursor.getLong(DbStructureLesson.Columns.VIEWED_BY),
            passedBy = cursor.getLong(DbStructureLesson.Columns.PASSED_BY),
            voteDelta = cursor.getLong(DbStructureLesson.Columns.VOTE_DELTA),
            language = cursor.getString(DbStructureLesson.Columns.LANGUAGE),
            isPublic = cursor.getBoolean(DbStructureLesson.Columns.IS_PUBLIC),
            createDate = cursor.getDate(DbStructureLesson.Columns.CREATE_DATE),
            updateDate = cursor.getDate(DbStructureLesson.Columns.UPDATE_DATE),
            learnersGroup = cursor.getString(DbStructureLesson.Columns.LEARNERS_GROUP),
            teachersGroup = cursor.getString(DbStructureLesson.Columns.TEACHERS_GROUP),
            timeToComplete = cursor.getLong(DbStructureLesson.Columns.TIME_TO_COMPLETE)
        )

    public override fun getContentValues(lesson: Lesson): ContentValues {
        val values = ContentValues()

        values.put(DbStructureLesson.Columns.ID, lesson.id)
        values.put(DbStructureLesson.Columns.TITLE, lesson.title)
        values.put(DbStructureLesson.Columns.SLUG, lesson.slug)
        values.put(DbStructureLesson.Columns.COVER_URL, lesson.coverUrl)
        values.put(DbStructureLesson.Columns.COURSES, DbParseHelper.parseLongArrayToString(lesson.courses))
        values.put(DbStructureLesson.Columns.STEPS, DbParseHelper.parseLongArrayToString(lesson.steps))
        values.put(DbStructureLesson.Columns.ACTIONS, gson.toJson(lesson.actions))
        values.put(DbStructureLesson.Columns.IS_FEATURED, lesson.isFeatured)
        values.put(DbStructureLesson.Columns.PROGRESS, lesson.progress)
        values.put(DbStructureLesson.Columns.OWNER, lesson.owner)
        values.put(DbStructureLesson.Columns.SUBSCRIPTIONS, DbParseHelper.parseStringArrayToString(lesson.subscriptions))
        values.put(DbStructureLesson.Columns.VIEWED_BY, lesson.viewedBy)
        values.put(DbStructureLesson.Columns.PASSED_BY, lesson.passedBy)
        values.put(DbStructureLesson.Columns.VOTE_DELTA, lesson.voteDelta)
        values.put(DbStructureLesson.Columns.LANGUAGE, lesson.language)
        values.put(DbStructureLesson.Columns.IS_PUBLIC, lesson.isPublic)
        values.put(DbStructureLesson.Columns.CREATE_DATE, lesson.createDate?.time ?: -1)
        values.put(DbStructureLesson.Columns.UPDATE_DATE, lesson.updateDate?.time ?: -1)
        values.put(DbStructureLesson.Columns.LEARNERS_GROUP, lesson.learnersGroup)
        values.put(DbStructureLesson.Columns.TEACHERS_GROUP, lesson.teachersGroup)
        values.put(DbStructureLesson.Columns.TIME_TO_COMPLETE, lesson.timeToComplete)

        return values
    }
}
